src/crypto/hsm/nxp/PersistentStorageOperationalKeystoreHSM.cpp

```diff
@@ -0,0 +1,279 @@
+ *    limitations under the License.
+ */
+
+#include "PersistentStorageOperationalKeystoreHSM.h"
+#include "CHIPCryptoPALHsm_SE05X_utils.h"
+#include <crypto/OperationalKeystore.h>
+#include <crypto/hsm/CHIPCryptoPALHsm.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <lib/support/SafeInt.h>
+
+#if ENABLE_HSM_GENERATE_EC_KEY
+
+namespace chip {
+
+using namespace chip::Crypto;
+
+namespace {
+
+// Tags for our operational keypair storage.
+constexpr TLV::Tag kOpKeyVersionTag = TLV::ContextTag(0);
+constexpr TLV::Tag kOpKeyDataTag    = TLV::ContextTag(1);
+
+// If this version grows beyond UINT16_MAX, adjust OpKeypairTLVMaxSize
+// accordingly.
+constexpr uint16_t kOpKeyVersion = 1;
+
+constexpr size_t OpKeyTLVMaxSize()
+{
+    // Version and serialized key
+    return TLV::EstimateStructOverhead(sizeof(uint16_t), Crypto::P256SerializedKeypair::Capacity());
+}
+
+} // namespace
+
+#define MAX_KEYID_SLOTS_FOR_FABRICS 32
+#define FABRIC_SE05X_KEYID_START 0x7D100000
+
+struct keyidFabIdMapping_t
+{
+    uint32_t keyId;
+    FabricIndex fabricIndex;
+    bool isPending;
+    Crypto::P256KeypairHSM * pkeyPair;
+} keyidFabIdMapping[MAX_KEYID_SLOTS_FOR_FABRICS] = {
+    0,
+};
+
+uint8_t getEmpytSlotId()
+{
+    uint8_t i = 0;
+    while (i++ < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        if (keyidFabIdMapping[i].keyId == 0 && keyidFabIdMapping[i].isPending == 0)
+        {
+            break;
+        }
+    }
+    return i;
+}
+
+bool PersistentStorageOperationalKeystoreHSM::HasOpKeypairForFabric(FabricIndex fabricIndex) const
+{
+    uint8_t i = 0;
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), false);
+
+    // If there was a pending keypair, then there's really a usable key
+    if (mIsPendingKeypairActive && (fabricIndex == mPendingFabricIndex) && (mPendingKeypair != nullptr))
+    {
+        ChipLogProgress(Crypto, "SE05x: HasOpKeypairForFabric ==> mPendingKeypair found");
+        return true;
+    }
+
+    while (i++ < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        if (keyidFabIdMapping[i].fabricIndex == fabricIndex)
+        {
+            ChipLogProgress(Crypto, "SE05x: HasOpKeypairForFabric ==> stored keyPair found");
+            return true;
+        }
+    }
+
+    ChipLogProgress(Crypto, "SE05x: HasOpKeypairForFabric ==> No key found");
+    return false;
+}
+
+CHIP_ERROR PersistentStorageOperationalKeystoreHSM::NewOpKeypairForFabric(FabricIndex fabricIndex,
+                                                                          MutableByteSpan & outCertificateSigningRequest)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t slotId = getEmpytSlotId();
+
+    VerifyOrReturnError(slotId != 0, CHIP_ERROR_NO_MEMORY);
```

Comment:
This does not make sense.  getEmptySlotId() never returns 0.

Reply:
Fixed this. Changed to 'VerifyOrReturnError(slotId < MAX_KEYID_SLOTS_FOR_FABRICS, CHIP_ERROR_NO_MEMORY);' now.